ast-checks/src/missing-patch-comment.rs

```diff
@@ -0,0 +1,102 @@
+use codespan::{FileId, Files};
+use comment_finders::{find_comment_above, find_comment_within};
+use rnix::types::*;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, env, error::Error, fs, process};
+use tree_utils::walk_keyvalues_filter_key;
+
+mod comment_finders;
+mod tree_utils;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct SourceLocation {
+    column: usize,
+    line: usize,
+    file: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct MissingCommentMessage {
+    cond: bool,
+    locations: Vec<SourceLocation>,
+    msg: &'static str,
+    name: &'static str,
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let args: Vec<String> = env::args().skip(1).collect();
+    let mut report = HashMap::new();
+
+    let mut files = Files::new();
+    for filename in args {
+        let file_id = files.add(filename.clone(), fs::read_to_string(&filename)?);
+        report.insert(filename.clone(), analyze_single_file(&files, file_id)?);
+    }
+
+    println!("{}", serde_json::to_string(&report)?);
+    Ok(())
+}
+
+fn analyze_single_file(
+    files: &Files<String>,
+    file_id: FileId,
+) -> Result<Vec<MissingCommentMessage>, Box<dyn Error>> {
+    let ast = rnix::parse(files.source(file_id)).as_result()?;
+    let root = ast.root().inner().ok_or("No elements")?;
+    let mut report: Vec<MissingCommentMessage> = vec![];
+
+    // Find a keys in the attrset called 'patches' that
+    // *do not* have a comment directly above them
+    let patches_without_top_comment = walk_keyvalues_filter_key(&root, "patches").filter(|kv| {
+        // Look for a comment directly above the `patches = ...` line
+        // (I'm interpreting that as the comment which applies to all patches.)
+        find_comment_above(kv.node()).is_none()
+    });
+
+    for kv in patches_without_top_comment {
+        // Attempt to cast the value of the `patches={value}` to a List
+        let value = match kv.value().and_then(List::cast) {
+            Some(l) => l,
+            None => {
+                let offset = kv.node().text_range().start().to_usize() as u32;
+                let loc = files.location(file_id, offset)?;
+                eprintln!(
```

Comment:
Could we make this produce a `Message` like any other error? Otherwise it will just crash:

```
Unexpected format at line 8 column 2. `patches` is not a list
Traceback (most recent call last):
  File "/nix/store/h8zbyy41jigbikyj9paiqyksppr3nbzp-nixpkgs-hammer/bin/.nixpkgs-hammer-wrapped", line 264, in <module>
    main(args)
  File "/nix/store/h8zbyy41jigbikyj9paiqyksppr3nbzp-nixpkgs-hammer/bin/.nixpkgs-hammer-wrapped", line 203, in main
    all_ast_check_messages = ast_checks(args.nix_file, args.attr_paths, args.excluded_rules)
  File "/nix/store/h8zbyy41jigbikyj9paiqyksppr3nbzp-nixpkgs-hammer/bin/.nixpkgs-hammer-wrapped", line 42, in ast_checks
    proc = subprocess.run(cmd, stdout=subprocess.PIPE, check=True, text=True)
  File "/nix/store/wkw6fsjasr7jbbrlakxxpbiapa8hws42-python3-3.8.7/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['missing-patch-comment', '/home/jtojnar/Projects/nixpkgs-hammering/tests/missing-patch-comment/comment-above.nix']' returned non-zero exit status 1.

```